File: src/main/java/alfheim/common/item/equipment/tool/ItemRealitySword.java
package alfheim.common.item.equipment.tool;

import java.nio.charset.Charset;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.List;

import javax.xml.bind.annotation.adapters.HexBinaryAdapter;

import akka.routing.RemoveRoutee;
import alexsocol.asjlib.ASJUtilities;
import alfheim.AlfheimCore;
import alfheim.api.AlfheimAPI;
import alfheim.api.ModInfo;
import alfheim.api.spell.DamageSourceSpell;
import alfheim.common.core.registry.AlfheimItems;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemSword;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.DamageSource;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.IIcon;
import net.minecraft.util.StatCollector;
import net.minecraft.util.Vec3;
import net.minecraft.world.World;
import vazkii.botania.api.mana.IManaUsingItem;
import vazkii.botania.api.mana.ManaItemHandler;
import static vazkii.botania.common.core.helper.ItemNBTHelper.*;

public class ItemRealitySword extends ItemSword implements IManaUsingItem {

	public static final String TAG_ELEMENT = "element";
	public static IIcon[] textures = new IIcon[6];
	
	public ItemRealitySword() {
		super(AlfheimAPI.REALITY);
		this.setCreativeTab(AlfheimCore.alfheimTab);
		this.setNoRepair();
		this.setUnlocalizedName("RealitySword");
	}
	
	@Override
	public String getUnlocalizedName(ItemStack stack) {
		return "item.RealitySword" + getInt(stack, TAG_ELEMENT, 0);
	}
	
	@Override
	@SideOnly(Side.CLIENT)
	public void registerIcons(IIconRegister reg) {
		for (int i = 0; i < 6; i++)
		textures[i] = reg.registerIcon(ModInfo.MODID + ":RealitySword" + i);
	}
	
	@Override
	@SideOnly(Side.CLIENT)
	public IIcon getIconIndex(ItemStack stack) {
		return textures[getInt(stack, TAG_ELEMENT, 0)];
	}

	@Override
	public IIcon getIcon(ItemStack stack, int pass) {
		return getIconIndex(stack);
	}
	
	public ItemStack onItemRightClick(ItemStack stack, World world, EntityPlayer player) {
		if (player.isSneaking()) {
			if (getInt(stack, TAG_ELEMENT, 0) == 5) return stack;
			if (merge(player.getCommandSenderName(), stack.getDisplayName()).equals("35E07445CBB8B10F7173F6AD6C1E29E9A66565F86AFF61ACADA750D443BFF7B0")) {
				setInt(stack, TAG_ELEMENT, 5);
				stack.getTagCompound().removeTag("display");
				return stack;
			}
			
			if (!ManaItemHandler.requestManaExact(stack, player, 1, !world.isRemote)) return stack;
			setInt(stack, TAG_ELEMENT, Math.max(0, getInt(stack, TAG_ELEMENT, 0) + 1) % 5);
		} else player.setItemInUse(stack, this.getMaxItemUseDuration(stack));
		return stack;
	}
	
	String merge(String s1, String s2) {
		String s = "";
		for (int i = 0; i < s1.length(); i++) for (int j = 0; j < s2.length(); j++) s += (char) ((s1.charAt(i) * s2.charAt(j)) % 256);
		return hash(s);
	}
	
	String hash(String str) {
		if(str != null)
			try {
				MessageDigest md = MessageDigest.getInstance("SHA-256");
				return new HexBinaryAdapter().marshal(md.digest(salt(str).getBytes(Charset.forName("UTF-8"))));
			} catch (NoSuchAlgorithmException e) {
				e.printStackTrace();
			}
		return "";
	}

	// Might as well be called sugar given it's not secure at all :D
	String salt(String str) {
		str = str += "wellithoughtthatthisiscoolideaandicanmakesomethinglikethis#whynot";
		SecureRandom rand = new SecureRandom(str.getBytes(Charset.forName("UTF-8")));
		int l = str.length();
		int steps = rand.nextInt(l);
		char[] chrs = str.toCharArray();
		for(int i = 0; i < steps; i++) {
			int indA = rand.nextInt(l);
			int indB;
			do {
				indB = rand.nextInt(l);
			} while(indB == indA);
			char c = (char) (chrs[indA] ^ chrs[indB]);
			chrs[indA] = c;
		}

		return String.copyValueOf(chrs);
	}
	
	@Override
	public void onUpdate(ItemStack stack, World world, Entity entity, int slotID, boolean inHand) {
		if (world.isRemote) return;
		
		boolean flag = getInt(stack, TAG_ELEMENT, 0) == 5;
		if (entity instanceof EntityPlayer) {
			if (!flag && 0 < getInt(stack, TAG_ELEMENT, 0) && getInt(stack, TAG_ELEMENT, 0) < 5) 
				if (!ManaItemHandler.requestManaExact(stack, (EntityPlayer) entity, 1, !world.isRemote)) 
					setInt(stack, TAG_ELEMENT, 0);
			
			if (flag && !entity.getCommandSenderName().equals("AlexSocol")) {
				EntityPlayer player = ((EntityPlayer) entity);
				world.spawnEntityInWorld(new EntityItem(world, player.posX, player.posY, player.posZ, stack.copy()));
				player.inventory.consumeInventoryItem(AlfheimItems.realitySword);
				player.setHealth(0);
				player.onDeath(DamageSource.outOfWorld);
				ASJUtilities.sayToAllOnline(StatCollector.translateToLocalFormatted("item.RealitySword.DIE", player.getCommandSenderName())); 
			}
		} else if (flag && entity instanceof EntityLivingBase) {
			EntityLivingBase living = ((EntityLivingBase) entity);
			world.spawnEntityInWorld(new EntityItem(world, living.posX, living.posY, living.posZ, stack.copy()));
			stack.stackSize = 0;
			living.setHealth(0);
			living.onDeath(DamageSource.outOfWorld);
		} else if (flag) {
			world.spawnEntityInWorld(new EntityItem(world, 0, 666, 0, stack.copy()));
			stack.stackSize = 0;
		}
	}
	
	@Override
	public boolean hitEntity(ItemStack stack, EntityLivingBase target, EntityLivingBase attacker) {
		if (attacker instanceof EntityPlayer) {
			int elem = getInt(stack, TAG_ELEMENT, 0);
			if (elem != 0 && (elem == 5 || ManaItemHandler.requestManaExact(stack, (EntityPlayer) attacker, 1000, !attacker.worldObj.isRemote && elem != 5))) useAbility(elem, attacker, target);
		}
		return super.hitEntity(stack, target, attacker);
	}
	
	private void useAbility(int i, EntityLivingBase attacker, EntityLivingBase target) {
		switch (i) {
			case 1: {
				Vec3 vec = attacker.getLookVec();
				target.motionX = vec.xCoord * 1.5;
				target.motionZ = vec.zCoord * 1.5;
				break;
			}
			case 2: target.addPotionEffect(new PotionEffect(Potion.poison.id, 80, 1)); break;
			case 3: target.setFire(6); break;
			case 4: {
				target.addPotionEffect(new PotionEffect(Potion.blindness.id, 100, -1));
				target.hurtResistantTime = target.hurtTime = 0;
				target.attackEntityFrom(DamageSourceSpell.water(attacker), 5);
				target.hurtResistantTime = target.maxHurtResistantTime;
				target.hurtTime = target.maxHurtTime = 10;
				break;
			}
			case 5: {
				for (int a = 1; a < 5; a++) useAbility(a, attacker, target);
				break;
			}
		}
	}
	
	@Override
	public void addInformation(ItemStack stack, EntityPlayer player, List list, boolean b) {
		int elem = getInt(stack, TAG_ELEMENT, 0);
		if (elem == 5) {
			list.add(StatCollector.translateToLocal("item.RealitySword.descZ"));
			return;
		}
		
		if (1 > elem || elem > 4) list.add(StatCollector.translateToLocal("item.RealitySword.desc0"));
		switch (elem) {
			case 1: list.add(StatCollector.translateToLocal("item.RealitySword.desc1")); break;
			case 2: list.add(StatCollector.translateToLocal("item.RealitySword.desc2")); break;
			case 3: list.add(StatCollector.translateToLocal("item.RealitySword.desc3")); break;
			case 4: list.add(StatCollector.translateToLocal("item.RealitySword.desc4")); break;
		}
		if (0 < elem && elem < 5) list.add(StatCollector.translateToLocal("item.RealitySword.desc5"));
	}

	@Override
	public boolean usesMana(ItemStack stack) {
		return (0 < getInt(stack, TAG_ELEMENT, 0) && getInt(stack, TAG_ELEMENT, 0) < 5);
	}
}

File: src/main/java/alfheim/common/item/equipment/tool/ItemRealitySword.kt
package alfheim.common.item.equipment.tool

import alexsocol.asjlib.*
import alfheim.api.*
import alfheim.client.render.world.VisualEffectHandlerClient.VisualEffects
import alfheim.common.core.handler.VisualEffectHandler
import alfheim.common.core.helper.*
import alfheim.common.core.util.AlfheimTab
import cpw.mods.fml.common.registry.GameRegistry
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.*
import net.minecraft.entity.boss.*
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.*
import net.minecraft.potion.*
import net.minecraft.stats.*
import net.minecraft.util.*
import net.minecraft.world.World
import vazkii.botania.api.mana.*
import vazkii.botania.common.core.helper.ItemNBTHelper.*
import kotlin.math.*

class ItemRealitySword: ItemSword(AlfheimAPI.mauftriumToolmaterial), IManaUsingItem {
	
	init {
		creativeTab = AlfheimTab
		setNoRepair()
		unlocalizedName = "RealitySword"
	}
	
	override fun setUnlocalizedName(name: String): Item {
		GameRegistry.registerItem(this, name)
		return super.setUnlocalizedName(name)
	}
	
	private var ItemStack.element
		get() = getInt(this, TAG_ELEMENT, 0)
		set(value) = setInt(this, TAG_ELEMENT, value)
	
	override fun registerIcons(reg: IIconRegister) {
		textures = Array(6) { reg.registerIcon(ModInfo.MODID + ":RealitySword$it") }
	}
	
	override fun getRarity(stack: ItemStack) = AlfheimAPI.mauftriumRarity
	
	override fun getIconIndex(stack: ItemStack) = textures.safeGet(stack.element)
	
	override fun getIcon(stack: ItemStack, pass: Int) = getIconIndex(stack)
	
	override fun onItemRightClick(stack: ItemStack, world: World, player: EntityPlayer): ItemStack {
		if (!player.isSneaking) {
			player.setItemInUse(stack, getMaxItemUseDuration(stack))
			return stack
		}
		
		if (stack.element == 5) return stack
		
		if (merge(ASJUtilities.mapGetKeyOrDefault(ContributorsPrivacyHelper.contributors, player.commandSenderName, "itsmemario"), stack.displayName) == "756179BA5B0697ED01B6CD292A3A726BACD5B99E0E624B37A11E18CE0B40B83E") {
			stack.element = 5
			stack.tagCompound.removeTag("display")
			return stack
		}
		
		if (!ManaItemHandler.requestManaExact(stack, player, 10, false)) return stack
		stack.element = (stack.element + 1) % 5
		
		return stack
	}
	
	fun merge(s1: String, s2: String): String {
		val s = StringBuilder()
		for (c1 in s1) for (c2 in s2) s.append(((c1.code * c2.code) % 256).toChar())
		return HashHelper.hash("$s", "wellithoughtthatthisiscoolideaandicanmakesomethinglikethis#whynot")
	}
	
	override fun onUpdate(stack: ItemStack, world: World, entity: Entity, slotID: Int, inHand: Boolean) {
		if (world.isRemote) return
		
		val safe = stack.element != 5
		
		if (entity is EntityPlayer && stack.element in 1..4 && !ManaItemHandler.requestManaExact(stack, entity, 10, !world.isRemote))
			stack.element = 0
		
		if (safe) return
		
		if (entity is EntityLivingBase) {
			if (entity is EntityPlayer)
				if(ContributorsPrivacyHelper.isCorrect(entity.commandSenderName, "AlexSocol"))
					return
				else
					ASJUtilities.sayToAllOnline("item.RealitySword.DIE", entity.getCommandSenderName())
			
			entity.health = 0f
			entity.onDeath(DamageSource.outOfWorld)
		}
		
		EntityItem(world, entity.posX, entity.posY, entity.posZ, stack.copy()).spawn()
		stack.stackSize = 0
	}
	
	override fun onLeftClickEntity(stack: ItemStack, player: EntityPlayer, entity: Entity): Boolean {
		leftClickEntity(stack, player, entity)
		return true
	}
	
	// all this just to set damage element to water -_-
	fun leftClickEntity(stack: ItemStack, player: EntityPlayer, entity: Entity) {
		if (!entity.canAttackWithItem()) return
		if (entity.hitByEntity(player)) return
		
		val elem = stack.element
		
		var damage = player.getEntityAttribute(SharedMonsterAttributes.attackDamage).attributeValue.F
		var knockback = 0
		var addDamage = 0f
		
		if (entity is EntityLivingBase) {
			addDamage = EnchantmentHelper.getEnchantmentModifierLiving(player, entity)
			knockback += EnchantmentHelper.getKnockbackModifier(player, entity)
		}
		
		if (!(damage > 0f || addDamage > 0f)) return
		
		val crit = player.fallDistance > 0f && !player.onGround && !player.isOnLadder && !player.isInWater && !player.isPotionActive(Potion.blindness) && player.ridingEntity == null && entity is EntityLivingBase
		if (crit) damage *= 1.5f
		damage += addDamage
		
		val src = DamageSource.causePlayerDamage(player)
		
		if (elem == 1 || elem == 5) src.setDamageBypassesArmor().setTo(ElementalDamage.AIR)
		if (elem == 2 || elem == 5) src.setMagicDamage().setTo(ElementalDamage.EARTH)
		if (elem == 3 || elem == 5) {
			if (!entity.isImmuneToFire && (entity !is EntityLivingBase || !entity.isPotionActive(Potion.fireResistance))) src.setFireDamage()
			src.setTo(ElementalDamage.FIRE)
		}
		if (elem == 4 || elem == 5) src.setDamageIsAbsolute().setTo(ElementalDamage.WATER)
		
		val succ = entity.attackEntityFrom(src, damage)
		
		if (!succ) return
		
		var fire = EnchantmentHelper.getFireAspectModifier(player) * 4
		
		if (elem == 3 || elem == 5)
			fire += 6
		
		entity.setFire(fire)
		
		if (player.isSprinting)
			++knockback
		
		if (elem == 1 || elem == 5)
			if (entity !is IBossDisplayData) knockback += 3
		
		if (knockback > 0) {
			entity.addVelocity((-sin(player.rotationYaw * Math.PI / 180) * knockback * 0.5), 0.1, (cos(player.rotationYaw * Math.PI / 180) * knockback * 0.5))
			player.motionX *= 0.6
			player.motionZ *= 0.6
			player.isSprinting = false
		}
		
		if (elem == 4 || elem == 5) {
			if (entity !is IBossDisplayData)
				entity.motionY += 0.825
			
			VisualEffectHandler.sendPacket(VisualEffects.SPLASH, entity)
		}
		
		if (crit) player.onCriticalHit(entity)
		if (addDamage > 0f) player.onEnchantmentCritical(entity)
		if (damage >= 18f) player.triggerAchievement(AchievementList.overkill)
		
		player.setLastAttacker(entity)
		
		if (entity is EntityLivingBase) EnchantmentHelper.func_151384_a(entity, player)
		
		EnchantmentHelper.func_151385_b(player, entity)
		
		player.addExhaustion(0.3f)
		
		var target = entity
		if (entity is EntityDragonPart && entity.entityDragonObj is Entity)
			target = entity.entityDragonObj as Entity
		
		if (target !is EntityLivingBase) return
		
		if (elem == 1 || elem == 5)
			target.addPotionEffect(PotionEffect(Potion.blindness.id, 100))
		
		if (elem == 2 || elem == 5)
			target.addPotionEffect(PotionEffect(Potion.moveSlowdown.id, 100, 1))
		
		stack.hitEntity(target, player)
		if (stack.stackSize <= 0)
			player.destroyCurrentEquippedItem()
		
		player.addStat(StatList.damageDealtStat, (damage * 10f).roundToInt())
	}
	
	override fun addInformation(stack: ItemStack, player: EntityPlayer?, list: MutableList<Any?>, b: Boolean) {
		val elem = stack.element
		
		addStringToTooltip(list, "item.RealitySword.abil$elem")
		
		if (elem == 5) {
			addStringToTooltip(list, "item.RealitySword.descX")
			return
		}
		
		addStringToTooltip(list, "item.RealitySword.desc$elem")
		
		if (elem in 1..4) addStringToTooltip(list, "item.RealitySword.desc5")
	}
	
	override fun usesMana(stack: ItemStack) = stack.element in 1..4
	
	companion object {
		
		const val TAG_ELEMENT = "element"
		lateinit var textures: Array<IIcon>
	}
}